### ISSUE TITLE
`mcell_mc_split_filt` followed by `mcell_mc2d_force_knn` : length mismatch errror

### ISSUE THREAD

**Issue by nixstix** [tanaylab/metacell#41]
Hi Tanay lab,

When running `mcell_mc_split_filt` function, followed immediately by `mcell_mc2d_force_knn`, we get the following error:

Error in mc2d_comp_cell_coord(mc_id, graph_id, mgraph, mc_xy, symmetrize = symmetrize) : 
  MC-ERR: Length mismatch in number of projected MC and overal mc

Are we missing a step in the pipeline to be able to correct the mismatch?

Thanks for any help, and thats for the package!

**Comment by akhiadber** [comment]
please paste the output of `length(mc@colors)` and `max(mc@mc)`, assuming your metacell object is named 'mc'. Assuming this is the mismatch, we'll look into why this mismatch happens when running `mcell_mc_split_filt`.

**Comment by nixstix** [comment]
Hi, thanks for the quick response! Indeed this was the mismatch. The output `length(mc@colors)` was 1, and `max(mc@mc)` was 22. When I populated mc@colors with 22 colors (e.g. using `mc_colorize_default` it worked.

**Comment by akhiadber** [comment]
We'll still look into why the length of mc@colors was 1 when running the split function, assuming you didn't set this on purpose. For now coloring is some manner is the workaround

**Comment by nixstix** [comment]
We did not set anything regarding the colours. If you need a reproducible example, I can email you a subset of data and the code we are using.

**Comment by XiangyangWang296** [comment]
> Hi, thanks for the quick response! Indeed this was the mismatch. The output `length(mc@colors)` was 1, and `max(mc@mc)` was 22. When I populated mc@colors with 22 colors (e.g. using `mc_colorize_default` it worked.

Indeed, I have met the similar error as above. After checking [mc_outliers.r](https://github.com/tanaylab/metacell/blob/HEAD/R/mc_outliers.r), here may be the issue:
`mcell_mc_split_filt = function(new_mc_id, mc_id, mat_id, T_lfc, plot_mats=T, dirichlet=F)
{
        #......
        new_mc = rep(NA, length(mc@mc))
	names(new_mc) = names(mc@mc)
        #......	
        new_mc = new_mc[setdiff(names(new_mc),new_outliers)]
	scdb_add_mc(new_mc_id, tgMCCov(new_mc, new_outliers, mat))
}`
Since new_mc@colors was no assigned, what I'll add to after running `mcell_mc_split_filt` is as followed:
```
mcell_mc_split_filt(new_mc_id=paste0(id,'_filt'),   
            mc_id=id, 
            mat_id=id,
            T_lfc=3, plot_mats=F)
# restore mc@colors as one unknown bug-> return 1 after function 'mcell_mc_split_filt'
mc_filt = scdb_mc(paste0(id,'_filt'))
mc_filt@colors = rep(NA, max(mc_filt@mc))
names(mc_filt@colors) = 1:max(mc_filt@mc)
scdb_add_mc(paste0(id,'_filt'),mc_filt)
```
That takes a fair bit of redundancy as it runs after 'mcell_mc_split_filt' but not in it. 

Hope that helps.
Best
